lib/buffer.dart

```diff
@@ -166,23 +166,23 @@ class ByteDataWriter {
   int bufferLength;
   final Endian endian;
   final _bb = BytesBuffer();
-  ByteData _data;
+  ByteData? _data;
   int _offset = 0;
 
   ByteDataWriter({this.bufferLength = 128, this.endian = Endian.big});
 
   void _flush() {
     if (_data != null) {
       if (_offset > 0) {
-        _bb.add(_data.buffer.asUint8List(0, _offset));
+        _bb.add(_data!.buffer.asUint8List(0, _offset));
       }
       _data = null;
       _offset = 0;
     }
   }
 
   void _init(int required) {
-    if (_data == null || _offset + required > _data.lengthInBytes) {
+    if (_data == null || _offset + required > _data!.lengthInBytes) {
       _flush();
       _data = ByteData(bufferLength > required ? bufferLength : required);
```

Comment:
The reason for setting the `_data` to `null` is this line: we can allocate more bytes than the default `bufferLength` when it is needed. To keep this efficiency, we could have an extra parameter on `_flush`, to initialize a new buffer with a non-default length - when needed.

Another -possibly short term- option is to rename `_data` to `_nullableData` and use `ByteData get _data => _nullableData!;` (not sure how this pans out, but may reduce the number of asserts in the code.)

Reply:
Yeah, I saw that. Essentially you need some sort of flag for this case. To keep things as similar as possible to what they used to be, I made a getter for data which is non-nullable, and a setter that is nullable but only sets a flag when a null is passed. Let me know if it looks like a good approach, or if I missed something.